### ISSUE TITLE
resource signal : have a way to make value() to persists while loading

### ISSUE THREAD

**Issue by sharakouze** [angular/angular#58602]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

My typical use case is a resource loading data for a sortable table like this :

```typescript
readonly tableResource = resource({
    request: () => ({
      sort: this.sort(),
    }),
    loader: async ({ request }) => {
      const { sort } = request;
      const response = await fetch(`http://localhost/api/test?sort=${sort}`);
      return await response.json() as TableData[];
  }
});
```

On the first execution when the data is loaded, _tableResource.value()_ contains my initial data.

When the user sort the table, the _sort_ signal is changed therefore the tableResource reloads the data. 

While the loader function is running, _tableResource.value()_ is set to **undefined**, then when the data is loaded _tableResource.value()_ contains my new sorted data.

I think this is the expected behavior but I have 2 problem with this :

1/ If the fetch function is slow to return a value, the user will see all the table content disappearing (because the resource value changes to undefined) before it shows again.
I think its best if the "old" data stays on the screen while it's loading (even if there is a loading indicator)

2/ If I want to use a linkedSignal with the resource value, it will be called twice for only one reload

### Proposed solution

I think It would be nice if there was a way to force value to persists while it's loading (like when we call the reload() function)

### Alternatives considered

Testing status() === 'ResourceStatus.Loading' and storing previous value outside resource

**Comment by wartab** [comment]
It was announced that there will be an RFC regarding resource soon, I think until then suggestions aren't being addressed.

**Comment by mauriziocescon** [comment]
Yep, I guess it would be a very welcome addition with `value` available in the loader as well:

```ts
  something = rxResource({
    request: () => this.params(),
    loader: ({ request, previous }) => {

      // value available together with the status
      return ... // something that can depend on value
    },
  });
```

Some extra examples: 

1) **Any sort of pagination**: while you're moving from page 1 to 2 and refetch data, you might want to still render the same value... and maybe just disable user interaction based on `isLoading`. 

2) **Infinite loading**: having the previous value available would make "concatenation" of different resource chunks straightforward (at least in some simple cases). There wouldn't be any need for an external "accumulator" built on top of the resource... assuming a single resource cannot / shouldn't be loaded in one go. 

3) **Signals composition**: at the end of the day, resources can also be seen as simple signals to be composed in order to get what's needed. From this perspective, the fact values turns undefined when params changes might be an undesirable effect cause it might generate unwanted changes / unwanted renderings / glitches / ... or simply extra code. 

My 2 cents!

**Comment by JeanMeche** [comment]
We discussed that topic yesterday and we see a value (pun intended) in keeping the value while reloading. The status / `isLoading()` could be used to determine if an async task is in progress. 

We want to give us a bit of time to make any change and gather an extended feedback. 

One of the question that arose from the discussion was, what usecase could take an advantage of setting the value to `undefined` while  reloading ?

**Comment by arturober** [comment]
This is another proposal to keep the loader function cleaner. This would be the use case for implementing pagination with infinite scrolling:

```typescript
  productsResource = rxResource<Product[], number>({
    request: () => this.page(),
    loader: ({ request: page }) => this.#productsService.getProducts(page),
    updateValue: (prev, current, {request:page}) => page === 1 ? current : prev.concat(current),
  });
```

By default the updateValue function would just return the current value.

**Comment by cyrilletuzi** [comment]
Note that it is not just a DX and UX issue, it is more importantly a performance one too. Let us take the table example again: as it will be value > undefined > new value, the table will be constructed with the first value, then unconstructed, then re-constructed from scratch for the new value, even if the table is done correctly by tracking a unique identifier. So it totally breaks optimization, which is a serious issue for tables with a lot of data.

> One of the question that arose from the discussion was, what usecase could take an advantage of setting the value to undefined while reloading ?

And I think this answer your question @JeanMeche: it should never be undefined while reloading, because it breaks the data-binding logic / flow.

**Comment by e-oz** [comment]
If a flag is set to keep the value (or if it’s the default option), should the value still be kept if the fetching request has failed?

**Comment by mauriziocescon** [comment]
Well, as mentioned in the ticket you can combine `resource` + `linkedSignal` to avoid rendering problems. It's just terribly annoying IMO! 

And yes: IMO the value should be kept even in case of errors. You can always use `status` / `error` to conditionally react to the failure. 

IMHO, there should be a new method called `reset` (or similar) that will set `value` to undefined / `defaultValue` (+ `status` reset) while `reload` will just retrigger the `loader` with the current `request`. 

Note that in my mind, `defaultValue` !== `initialValue`: 
- `defaultValue`: value whenever reset is called or during the first loading, 
- `initialValue`: value at `resource` creation time (cause I already have the data): no initial loading; I might `reload` it in the future.

Assuming I'm not missing something stupid, that's more or less what I would do!

**Comment by JeanMeche** [comment]
Capturing some recent discussions we had with the team, and also what recently landed in the main branch (which will be reflected in the next pre-release). 

* defaultValue / initialValue are the same. 
* The resource resets to `defaultValue` when loading. 
* A resource represents a state, when you (re)load it should be obvious that there is no current state as it is currently being loaded.
  * But the previous value should remain accessible, but in a separate field. (This isn't implemented yet).

**Comment by mauriziocescon** [comment]
Ok, I see! 

Say, it's very likely just me... but I still find strange that value is turned to `defaultValue` during loading precisely because I consider `value` state. Changed asynchronously, but still state. 

In my mind, reading a loading `resource` is like reading a `computed` when the graph is dirty: I'm reading something is not yet finalised, so I'd expect to get the previous value until the `async set` is done - either with success or error. 

Of course, `status` / `error` are part of the equation and should be used to decide the nature of `value()` (is there a loading ongoing? was there an error?). 

Anyway: thanks for the feedback!

**Comment by mauriziocescon** [comment]
Quickly coming back to the discussion (I know... 😅) just to point out last Friday [stream from Ryan Carniato](https://www.youtube.com/watch?v=xnmvxWEK25I) regarding signals_v2 in solid. 

It's a very long one, but worth watching IMO... and in the first half, it touches this point as well (new solid `createAsync` + `effect` + flag). 

Here is an ng-adaption of one the examples discussed: an unavoidable waterfall case where a child component depends on a parent `resource.value` with a straightforward implementation.  

```ts
// Parent feeds Child with a number; asyncOperation built to magnify problems

// just returning value * 2 after time ms
function asyncOperation(value: number, time: number): Promise<number> {...}

@Component({
  template: `
    <div>Child value: {{ r.value() }}</div>
  `,
})
export class Child {
  source = input.required<number>();

  r = resource({
    request: () => ({ source: this.source() }),
    loader: ({ request }) => asyncOperation(request.source, Math.random() < 0.5 ? 1000 : 2000),
  });

  log = effect(() => console.log(`Parent r.value: ${this.source()}, Child r.value: ${this.r.value()}`));
}

@Component({
  template: `
    <child [source]="r.value()"/>
  `,
})
export class Parent {
  source = signal(0);

  r = resource({
    request: () => ({ source: this.source() }),
    loader: ({ request }) => asyncOperation(request.source, 1500),
    // defaultValue: 0,
  });

  changeSource() {
    this.source.update(v => v + 1);
  }
}
```
https://stackblitz.com/edit/stackblitz-starters-fnqucjtx?file=src%2Fmain.ts

Despite being straightforward, this implementation is completely off: 
- it has a glitch on the child component, 
- which makes the output of log-effect unpredictable. 

Of course, you might say: just refactor the code, you can protect yourself from problems like this, you'll pass previous value on the Parent (when available), ... and so on. 

But IMHO it also clearly shows that having `resource.value` changing (while loading) can really compromise the code correctness in straightforward usages of `resource`! 

Again: considering `resource` a signal which has `resource.value` changed asynchronously only in case of "success" seems to me the most reasonable choice! Any other (more advanced) case can be covered by deriving from `isLoading`, `error`, `status`, ... 

My last 2 cents (I promise you)!

**Comment by mariojsnunes** [comment]
a hacky workaround where value keeps the value if next is undefined:
```
result = resource(...);
value = computed(() => this.result.value(), { equal: (_, b) => b === undefined });
```
if you want to clear the data, ensure the next value returns null.

**Comment by gabynevada** [comment]
I feel we can have better DX if the resource keeps it's current value while reloading or at least provide the option to.

For anyone else, found the RFC and the discussion around keeping the last value. https://github.com/angular/angular/discussions/60121#discussioncomment-12329768

Also found this nice workaround that I got from the discussion using a linkedSignal (https://github.com/angular/angular/discussions/60121#discussioncomment-12331951)

```ts
data = httpResource(...);
memoData = linkedSignal({
  source: this.data,
  computation: (data, previous) => previous?.value ?? data,
);
```

**Comment by DanielNovaesDias** [comment]
```Typescript

import {
    resource,
    linkedSignal,
    ResourceStatus,
    ResourceRef,
    PromiseResourceOptions,
} from '@angular/core';

export function preservedResource<T, R>(
    config: PromiseResourceOptions<T, R>
): ResourceRef<T> {
    const original = resource(config) as ResourceRef<T>;

    const preserved = linkedSignal<
        {
            value: T;
            status: ResourceStatus;
        },
        T
    >({
        source: () => ({
            value: original.value(),
            status: original.status(),
        }),
        computation: (current, previous) => {
            if (current.status === ResourceStatus.Loading && previous) {
                return previous.value;
            }
            return current.value;
        },
    });

    return {
        ...original,
        value: preserved,
    };
}

```

Here's a little helper/wrapper function so it's easier to use the linkedSignal approach, seems to work fine for me at least.

**Comment by StephenFluin** [comment]
+1 this is still needed and should be an option of httpResource (or the default). It's messy to create computed's or linkedSignals.

**Comment by davidbusuttil** [comment]
+1

**Comment by alebx** [comment]
I just stumbled upon this issue and created a helper function:

```ts
/**
 * Keep the old datum until the loading is done
 * @param ref
 */
export const resourceWithRetention = <T>(ref: ResourceRef<T | undefined>): ResourceRef<T | undefined> => {
  const value = linkedSignal<{ value: T; isLoading: boolean }, T>({
    source: () => ({ value: ref.value(), isLoading: ref.isLoading() }),
    computation: (newValue, previous) => (newValue?.isLoading ? previous?.value : newValue?.value),
  });

  return new Proxy(ref, {
    get(target, prop, receiver) {
      if (prop === 'value') {
        return value.asReadonly();
      }

      return Reflect.get(target, prop, receiver);
    },
  });
};
```

**Comment by janschaeppi** [comment]
+1
IMHO retaining the latest value should just be the default behavior of the `value` signal.
Using `@if(resource.isLoading()) { .. }` is very natural in case one prefers not to display the latest value while loading.

**Comment by skoropadas** [comment]
I believe the most flexible and intuitive approach would be the following behavior:

- `value()` - during the initial load, it holds the `defaultValue` (which might be better named `initialValue`), and subsequently returns the last successfully loaded value. Even if the resource enters an error state after calling `reload()`, `value()` should still return the last successfully loaded value.
- `isLoading()` - reflects the loading state.
- `error()` - stores the error until the first successful load and resets as soon as the data is successfully loaded.

This would cover all possible use cases and allow the state to be displayed according to business logic requirements. For example, if I only display one state, I can use it like this:

```html
@if (!r.loading() && !r.error()) {
  {{ r.value() }}
}

@if (!r.loading() && r.error()) {
  {{ r.error() }}
}

@if (r.isLoading()) {
  Loading...
}
```

Or, if the business logic requires displaying everything at once, I can use this more simply:

```html
{{ r.value() }}

@if (r.error()) {
  {{ r.error() }}
}

@if (r.isLoading()) {
  Loading...
}
```

Often, this logic is also wrapped in UI components that take these states as inputs, and such an API would make working with them easier.

```html
<ui-table [data]="s.value()" [isLoading]="r.isLoading()" [error]="r.error()" />
```

**Comment by wawyed** [comment]
Any updates on this? @JeanMeche

**Comment by magwo** [comment]
Also interested in updates on this @JeanMeche 

This seems like a very common use case in Angular apps that strive for a pleasant UX. The default and unmodifiable behaviour in Resources is putting a wet blanket on my organisation's migration to Resources. I work with a fairly large hybrid iOS/Android/Web Angular application with roughly 4000 component types. We generally love Signals but this Resource disappearing-value-behaviour is causing friction.

The current approach doesn't support all use cases - requires workarounds which worsens DX and succinctness.

I like the proposal by @skoropadas . With an API like that you cover all the use cases. The current approach, where the resource clears its value upon reload or error, makes some use cases more complicated.

**Comment by mauriziocescon** [comment]
Hey! 

I spent some time on `resource` during the Christmas break: after ~1 year, I still feel `resource` should match the set of features from solid [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource#the-fetcher$0) in order to empower its usage. 

In practice: 
1. value kept among params changes (the essence of this ticket: some sort of new entry called previousValue would be fine as well), 
2. value provided in the loader, 
3. possibility of passing an object in the reload function (monitoring / tracking),
4. possibility of defining an initial value. 

Just some thoughts after many attempts to use it for real in various ways (services, components, ...). And sure: we can always ask agents do create a more personalised version of it... but I'm not 100% sure it's desirable for a primitive. 😅

Cheers!


PS: a solid stackblitz showcasing the 4 points above
```ts
function delay(ms: number) {...}

function NumberGenerator() {
  const [source, setSource] = createSignal<number | undefined>(undefined);

  // 1) data() kept among source changes
  const [data, { refetch }] = createResource<string[], number>(
    source,

    // 2) value provided in the loader
    // 3) refetching: param from refetch(param)
    async (source: number, { value, refetching }) => {
      return new Promise(async (resolve, reject) => {
        const r = Math.round(Math.random() * 100) / 100;
        await delay(r * 2000);
        if (r < 0.75) {
          resolve([...(value ?? []), `${r} ${refetching ? refetching : ''}`]);
        } else {
          reject(new Error('random failure'));
        }
      });
    },

    // 4) can define an initial value
    { initialValue: [`0`, `1`] }
  );

  // ...

  return (
    <>
      ...
    </>
  );
}
```
https://stackblitz.com/edit/solidjs-templates-kdsa4w8f?file=src%2FApp.tsx

**Comment by magwo** [comment]
Just a question: It seems clear that many users want the value to be retained while **reloading** a resource.

However, do we **also** want to the value to be retained when the resource is loading another value, because a dependency (for example an id) changed value and triggered the resource to load again?

I think I want the value to be retained in both cases, and leave it up to consumers to base UI logic on isLoading(), hasError() or state etc.

**Comment by meriturva** [comment]
I'm currently forced to use a linkedSignal as a fallback. Despite the experimental status, is this request on the roadmap? I'm a bit unsure about the direction to take. This is my current solution for making rxResource work with Kendo UI, maybe could help someone:

```typescript

import { DataResult, DataSourceRequestState, toDataSourceRequestString } from '@progress/kendo-data-query';

public kendoGridResource(state: Signal<DataSourceRequestState>): { resource: ResourceRef<DataResult>, data: Signal<DataResult> } {
    const resource = rxResource({
      defaultValue: {
        data: [],
        total: 0
      },
      params: () => ({
        state: state()
      }),
      stream: ({ params }) =>
        this._httpClient
          .get<DataResult>(`${this.baseUrl}/blabla?${toDataSourceRequestString(params.state)}`)
    });

    const data = linkedSignal({
      source: () => ({
        value: resource.value(),
        status: resource.status(),
      }),
      computation: (source, previous) => {
        if (previous && source.status === 'loading') {
          return previous.value;
        }
        return source.value;
      }
    }) as Signal<DataResult>;

    return {
      resource,
      data
    }
  }
``` 
This workaround prevents the grid from flickering during pagination.

**Comment by JeanMeche** [comment]
#66328, adds support for resource composition and gives an example to persist the value of a resource on reload. 
This feature will be shipped in 21.2 later this month.

**Comment by StephenFluin** [comment]
It feels like we still have to write a ton of boilerplate code in [the solution](https://github.com/angular/angular/pull/66328). Is there any world where we can get this as a built in type extending Resource, or as an exported helper method?

I know there's likely super valid reasons to not have this be the default, but I still can't think of any world where I want the value to disappear during a refresh when I can check that state elsewhere.

**Comment by mauriziocescon** [comment]
(+1)

It's also state... so I find it strange it's not managed directly by the framework (by default, as an option, ...). Otherwise I have the feeling we're back to third-party deps for doing fundamental "state related" things... with all the old problems (see observables and so on). 

I don't expect the ng-team to provide a sort of TanStack query: that would be too much... but immediately pulling a third party dep (or create wrappers) the moment I need to do "not trivial async operations" seems off to me. 

My 2 cents

**Comment by davidbusuttil** [comment]
(+1)

This should be an option instead of additional boilerplate.

**Comment by skoropadas** [comment]
If there is a super valid reason for the current Resource API, I’d like to hear it, because right now it looks as if the developers, instead of changing the existing API, are rolling out even more new features and utilities to solve problems created by previous solutions.

And I still absolutely don’t understand why all these utilities are needed if everything can be solved by changing the Resource API. Even if I want to reset the `value` on load, I can easily do that without any additional utilities

**Comment by magwo** [comment]
I agree, the snapshot solution seems a bit awkward and complicated compared to just changing the default to the most reasonable option for most users. I haven't seen anyone wish for the value to disappear during reload. The snapshot+linkedsignal is only marginally better than the Reflect/proxy based workaround suggested in this thread (which I'm currently using successfully).

Also, isn't a snapshot just the value of the signal at some point in time? Why do we need to convert into a "snapshot signal", when we can just let local variables be captured by a closure or similar? What are the advantages of the snapshot concept?

Maybe the snapshot composition possibilities are nice, but it doesn't really solve the problem that the Resource default behaviour isn't what users of Angular want/need, in general. So for example when we migrate code between libraries we need to also migrate the non-standard wrappers around resource like this `withPreviousValue`.

**Comment by magwo** [comment]
There's a [pretty long article about this](https://angular.schule/blog/2025-10-rx-resource-is-broken) (and two other issues) that describes this issue as a bug. I don't necessarily agree with that conclusion entirely, but the author provides very clear arguments for his viewpoint.

**Comment by austinw-fineart** [comment]
An entire year has gone by and I have yet to see one valid use case for undefined reload that isn't already covered by isLoading (i.e. I want to hide the stale value).

**Comment by sod** [comment]
I admire that @JeanMeche came up with something composable. It would be a shame if `httpResource` would grow into a kitchen sink of features thats not tree shakable.

**Comment by magwo-aza** [comment]
@sod it doesn't become a kitchen sink just because you change a behavior.

The problem is that no-one seems to need the behavior that `httpResource` currently exhibits. Like @austinw-fineart mentions, we have yet to see an example where retained value causes any issues.

Anyone who doesn't want to use retained values has multiple simple approaches:
1. Use the already-available isLoading(), errror() and state() helper signals.
or
2. Explicitly reset the value before refresh.

**Comment by arturober** [comment]
@magwo-aza I agree, the problem is the default behavior. It should be the opposite: keep the value by default and use composition functions to alter that (for example by setting undefined when loading, mapping the result automatically, etc.). Or at least add a simple option (boolean) to change this. I don't think that would be overkill either...

Please, consider this before making the API stable

**Comment by StephenFluin** [comment]
@sod often times making an object/method composable can mean larger bundle size to support the composability, even with tree shaking. It's definitely not a kitchen sink, I think we're all saying the default behavior is wrong.

**Comment by sod** [comment]
Sorry, my kitchen sink comment was derogatory. You have valid points. It is a cool feature to have baseline. Our app does exactly this since 2017 via rxjs and now via ngrx because it's exactly how UI should behave (fade out old data on load until refresh, not jank it out of the DOM while a request is being made).

All I wanted to say is, it's cool that there is a composable way. I never had a problem with bad or cumbersome APIs. I wrap it in my own `function ourResource` that applies the defaults that suit my needs. Our app needs tons of stuff that will never be built in.
* automatic retry on flaky network, but only on certain return codes
* throttle between requests
* automatic error handling for certain error codes
* automatic redirect if the firewall blocked a request

and much more.

So I'm not for or against any API changes here. Was just giving unsolicited context 🙈
